### ISSUE TITLE
SAML authentication back-end?

### ISSUE THREAD

**Issue by sochoa** [hashicorp/vault#370]
My team at work is wondering if this project has plans to support SAML, Github Enterprise (via OAuth), or both, anytime soon?  We have a need for something like Vault.

**Comment by sethvargo** [comment]
Hi @sochoa 

Thank you for opening an issue. This is not something that is currently planned, but sounds like a great idea. I have tagged as an enhancement. We would also gladly welcome a community contribution for this functionality.

**Comment by sochoa** [comment]
I'm game for contributing.  Does the project have any docs on writing an authentication back-end?

**Comment by sethvargo** [comment]
Hi @sochoa 

We do not have an "developer documentation", but you can follow the pattern in the existing GitHub authentication backend, for example. You might also be able to extend the existing GitHub backend to support a custom URL (and then everything else should "just work") :smile:

**Comment by armon** [comment]
@sochoa supporting GH enterprise is something we want to do, but SAML is the more questionable one as @sethvargo pointed out. I think supporting a custom OAuth endpoint in the existing provider should not be too hard! It may be worth splitting this ticket into two to clarify between the two.

**Comment by dalehamel** [comment]
@armon @sethvargo @sochoa 

I think SAML would be a huge win, as it is a standardized authentication interface, rather than having to implement something specific for each provider.

For instance, I'd like to use vault with Okta. I could write an integration specific ta Okta, but that seems like a waste since Okta just supports SAML.

I'll have to read up more on how SAML works to see if it's within my depth to contribute support for it, but I think that that this should really be a core feature of vault.

**Comment by jefferai** [comment]
For future reference, mattbarid/gosaml can handle client requests; it may be a useful starting point for anyone trying to code a server library. I do think that the core functionality for managing requests (parsing, signing, etc.) probably needs to be coded outside of Vault (perhaps as enhancements to gosaml) rather than living in a backend.

**Comment by jefferai** [comment]
Another thing to mention is that Vault may be able to act as a signer for SAML applications, when given an appropriate private key, but currently there is no way for Vault to handle raw XML requests (although it can generate them). This limits the utility of a Vault SAML backend right now since there would have to be an actual SAML server to handle requests from third-parties.

**Comment by andrewstuart** [comment]
I have the rough beginnings of a [SAML2 client](/andrewstuart/gosaml2) put together, which can decrypt but not yet authenticate (xmldsig and xml-c14n are seriously painful) encrypted SAML assertions from Shibboleth.  Thus far it only works for the specific settings Shibboleth was configured with (the defaults for [this docker container](/jtgasper3/docker-shibboleth-idp)), but those were at least pretty solid settings -- RSA-encrypted AES key for decrypting the main assertion body. 

I know there are quite a few caveats in there, but IIRC mattbaird's client doesn't do much in the way of SAML2 assertions, just in generating the Authentication Request. Kinda seems like it was abandoned pretty early on, unfortunately. I know I'd have a lot of interest in getting a proper Go SAML implementation working, and I'd love it if the work I put into that library ends up going somewhere.

**Comment by radeksimko** [comment]
SAML backend is on my radar too, although from what I found so far it doesn't look as easy as I originally thought.

Even though SAML is kind of standard protocol, it only defines the core part of the user flow, typically beginning and end (those are the parts covered by most SAML Go libraries you find on the internet today). Anything in between - i.e. forms & APIs & redirects is left for IdP's creativity (unfortunately).

Hence I think that proper Vault SAML auth backend will require not only [`mattbaird/gosaml`](https://github.com/mattbaird/gosaml) but also IdP specific logic, e.g.

 - ADFS - https://github.com/wernerb/aws-adfs
 - Ping Identity
 - OneLogin
 - Okta
 - Shibboleth
 - etc.

Some IdPs provide [APIs](https://developers.onelogin.com/api-docs/1/saml-assertions/generate-saml-assertion) which may make the implementation easier, but some just don't, so the library may just end up parsing form fields in HTML and sending `POST` requests to submit a form. :cry: 

Even though most of the IdP-specific and SAML-specific logic can (and should) be outside of Vault, I think that Vault will need to recognise the differences between IdPs in configuration.

**Comment by andrewstuart** [comment]
So I'd definitely like to revisit this, as I've made some leaps integrating the separate efforts of @russellhaering (and @phoebesimon) on [their saml2 library](/russellhaering/gosaml2) which did not support decryption but does validate signatures, and [my earlier effort](/andrewstuart/gosaml2), which can decrypt but not validate signatures. I think we'll have a solid foundation once russellhaering/gosaml2#5 is ready for merge.

Would there be any precedent for vault opening up a browser window for the authentication piece, redirecting to a known dynamic URL (to tie it to the shell authn request), then closing the browser window? My main concern is that the SAML flows I'm aware of will likely make console authentication a royal pain, and definitely not a great user experience.

**Comment by jefferai** [comment]
@andrewstuart Probably the right thing to do would be to have a separate application that uses Vault's API (the CLI is just an API client).

**Comment by radeksimko** [comment]
@jefferai what would be such application passing to the CLI/API? SAML assertion bodies?

**Comment by jefferai** [comment]
@radeksimko I don't know? I was just answering the question posed.

**Comment by andrewstuart** [comment]
Haha so the root of the issue is that SAML is built around the browser, and is asynchronous. Though I'm only familiar with Shibboleth, I think nearly all Identity Providers (IdPs) are browser-based for the AuthN step, and the Redirect (assertion as xml encoded in a base64 query param) or form POST bindings (assertion as xml encoded in a base64 POST body) directed at the Relying Party (aka Service Provider) assertion consumer endpoint.

So to have an intermediary application, it must be a long-lived well-known application because it must have its public certificate registered with the IdP and associated with the full URL of the assertion consumer. I don't see any particular problem with this, but at that point, it's more of a sidecar application than an authentication backend, as the certs/keys would most likely need to live in the sidecar. Vault would be pretty unaware of the AuthN/Z flow in this case.

So let me propose an Auth flow, just to flesh out my own thoughts and get some feedback.

1. A Vault API client initiates AuthN via SAML.
2. The Vault API generates a signed SAML AuthN Request URL, and returns the Request to the client along with the Request ID and/or a generated secret (and perhaps explicit expiration), as the Request ID may not be confidential. The client can use this Request ID and/or generated secret to query the status of the SAML AuthN request and obtain longer-lived creds upon successful SAML AuthN.
3. The Vault API client uses the SAML AuthN request however it sees fit.
  - For the official cli, I would imagine opening the URL in the default browser.
4. Vault receives the SAML assertion at an AssertionConsumer endpoint, and validates that the IdP is trusted, assertion is valid, etc. As the client of this endpoint is possibly a browser, return a message indicating status of the SAML Auth request, and stores the success for the ID/secret it generated earlier, with expiration at some point.
5. As this is an async flow, the Vault API client either polls or holds open a connection to the Vault API (identifying the "AuthN session" via the ID and/or secret it received) and, upon successful SAML AuthN, is granted a session token.

I'm sure I've missed a few things and it could likely be improved, but I definitely wanted to get something out there to start with. :-) Let me know what you think.

**Comment by argais** [comment]
are there any plans/roadmap/timeline/etc on this?

**Comment by jefferai** [comment]
Nothing currently, sorry.

**Comment by DarioAmiri** [comment]
This is profoundly disappointing. The issue is closed, but I can't help myself from addressing the incorrect assumptions/conclusions drawn by @radeksimko.

- SAML is not "kind of a standard" it is an exhaustively documented standard that is rigorously specific about all communication between the Identity Provider (IdP) and the Service Provider (SP), which would be Vault's role in this case.

- Besides the [SAML Core](http://www.oasis-open.org/committees/download.php/35711/sstc-saml-core-errata-2.0-wd-06-diff.pdf) specification, there is also the [SAML Binding](http://www.oasis-open.org/committees/download.php/35387/sstc-saml-bindings-errata-2.0-wd-05-diff.pdf) specification, which is rigorous about how to use POST and Redirect bindings to pass SAML messages between the SPs and IdPs.

- Furthermore, the [SAML Profile](http://www.oasis-open.org/committees/download.php/35389/sstc-saml-profiles-errata-2.0-wd-06-diff.pdf) specification provides all the details necessary for browser-based SAML SSO.

-  Yes, SAML does not say anything about how to create the login form on the IdP side, but this does not matter because it's _past the layer of abstraction_. The only thing a Vault SAML authn backend should concern itself with, is how to pass messages back and forth between the Vault and any given SAML IdP. How the IdP authenticates the user is completely irrelevant. In fact, that's actually the reason why you want to support SAML, because it allows the IdP to authenticate the user in whatever way it sees fit. I.e. showing a familiar corporate login form, using various additional factors of authentication, etc.

- Yes SAML is a difficult standard to implement and the Go libraries for SAML support are not mature. So consider implementing something like [OpenID Connect](http://openid.net/connect/) (OIDC) instead. OIDC has gained a fair amount of traction as a simpler alternative to SAML.

In conclusion, I hope the Vault project will reconsider/consider implementing a SAML or OIDC authentication backend. It would simplify integration and help drive adoption, especially among identity professionals who know and trust these standards. And, at the risk of being labeled a conspiracy theorist, _I really hope_ that the reason for not supporting these standards has nothing do with the fact that it would enable third-party or in-house IdPs to provide MFA, instead of having to buy Vault Enterprise to get this feature.

**Comment by jefferai** [comment]
@DarioAmiri the issue isn't closed...

**Comment by DarioAmiri** [comment]
@jefferai My bad. I think I got a little too fired up, I will curb my enthusiasm.

**Comment by jefferai** [comment]
@DarioAmiri I also want to point out that although @radeksimko does work at HashiCorp now, at the time of his comments he was not a HashiCorp employee, and he is not on the Vault team. They should not be considered to be HashiCorp's official position.

**Comment by jefferai** [comment]
I feel the need to clarify something:

@radeksimko is totally awesome. My previous comment is not a rebuke of his opinions, it is only meant to say that the Vault roadmap is set by the Vault team; what I was trying to get across is that while his technical input is welcomed (as is everybody's), it should not be viewed by others as a statement about the Vault roadmap.

Also, thanks for the apology -- for guidance around comms we often refer to our https://www.hashicorp.com/community-guidelines

**Comment by DarioAmiri** [comment]
@jefferai Thanks for clarifying. FWIW I have no reason to doubt @radeksimko's awesomeness. My intent was to highlight incorrect conclusions/information in the thread.

**Comment by getjack** [comment]
Uhm this issue is a bit confusing. It the ticket now closed, is it still open or in other words what is the current state on the issue?

**Comment by Oneiroi** [comment]
@getjack #370 is still open, the confusion came as other issues which reference this one are closed and users 'skim reading' has lead to confusion and presumption of this issue being closed.

I am unsure as to the hashicorp team's priorities on implementing this functionality, however afaict this issue is still very much open.

**Comment by jefferai** [comment]
It's still under consideration, just not a priority. Many, many more customers are asking us for OIDC rather than SAML so that's a higher priority.

**Comment by steven-cuthill-otm** [comment]
we are holding off implementing vault until we have SAML that will work with AzureAD with proper role assignment.

**Comment by steven-cuthill-otm** [comment]
4 years later ? any updates on this or any support for AAD auth / office 365  ?

**Comment by DarioAmiri** [comment]
IIRC HashiCorp sells SAML, and SSO support in general, as an enterprise feature. Therefore, it is unlikely that you will see support for this in the open source version.

**Comment by jefferai** [comment]
@DarioAmiri this is not correct. We do not support SAML in any edition of Vault right now, and our SSO support (in the form of OIDC) is in OSS.

**Comment by jefferai** [comment]
@steven-cuthill-otm Many of our customers are using OIDC auth with AAD.

**Comment by jtcressy** [comment]
I'll add my quick 2-cents. 

OIDC is a much nicer protocol than SAML, but at my company the powers-that-be decided not to include OIDC in purchasing our okta license. So, we can only use SAML for sso. It's dumb, and i'll be working to fix that in the future, but I'm sure other subscribers of okta may accidentally do this. After all, every other enterprise app supports saml, so the decision making process behind *only* purchasing the saml peice was that we would never really need to use oidc. 

I'm not implying this should be enough of a justification to implement saml in vault, but it's currently a pain point at my org. At the moment, anyway, vault is working just fine with the okta auth backend. I've just received feedback from many people that would like it if the auth flow used web redirects instead of users entering their passwords into a non-okta webpage (vault).

**Comment by rayterrill** [comment]
@steven-cuthill-otm I was able to get this working in AzureAD with OIDC, and the ability to assign role-based access in Vault based on AzureAD groups. It's definitely not super well documented IMO (I'll see about maybe adding some info to the docs), but it does indeed work.

**Comment by toddmichael** [comment]
> is there any integration with ADFS yet ?? SAML ??

@XinfinityoO if you're using ADFS 4.0 you can use the OIDC auth method which works quite nicely. If ADFS <= 3.0, I don't know of a way given absence of SAML backend.  It seems momentum these days is toward OIDC.

**Comment by XinfinityoO** [comment]
> > is there any integration with ADFS yet ?? SAML ??
> 
> @XinfinityoO if you're using ADFS 4.0 you can use the OIDC auth method which works quite nicely. If ADFS <= 3.0, I don't know of a way given absence of SAML backend. It seems momentum these days is toward OIDC.

Hi
Could you please direct me to a guide on how to do this. Vault has azure example but its different to Windows 2016 ADFS setup

Thanks

**Comment by Oneiroi** [comment]
@XinfinityoO https://support.vaultgrc.com/hc/en-us/articles/115006919687-Configuring-Vault-For-Single-Sign-On-ADFS first result on search, does this help ?

**Comment by XinfinityoO** [comment]
> @XinfinityoO https://support.vaultgrc.com/hc/en-us/articles/115006919687-Configuring-Vault-For-Single-Sign-On-ADFS first result on search, does this help ?

Not really, that is not hashicorp vault ..some other product

**Comment by monwolf** [comment]
Hi, 

After 5 years, I have a question: Is it planned to add this enhancement to your roadmap? 

Regards!

**Comment by trentondyck** [comment]
> Chiming in here. SAML would allow some large customers to integrate this.

This.  We're already configuring Jenkins, Bitbucket with SAML SSO by Azure AD.  Vault is the last remaining piece.

**Comment by pratiklotia** [comment]
+1 for SAML support. High time this feature is supported.

**Comment by mkrakowitzer** [comment]
After contacting hashicorp support, doesn't seem like this is on the roadmap for 2021. it's a blocker for us purchasing the enterprise version.

**Comment by jboero** [comment]
@radeksimko have you done anything more on this?  I've seen a few requests for it lately as well.  Happy to help any way I can.

**Comment by dosgeek** [comment]
+1 for SAML

**Comment by DemiMarie** [comment]
I am not a current or near-future user of Vault, but I can give some recommendations when it comes to implementing SAML.

SAML is SERIOUSLY difficult to implement correctly.  For one, Go’s built-in XML lexer and parser are not appropriate for the job.  My understanding SAML requires full XML standards compliance in just about every area except DTDs, even when given malicious input.  In particular, SAML requires that the library round-trip malicious XML successfully.  Go doesn’t make this guarantee.

From what I can tell, by far the safest way to implement XML-DSIG is something like (in pseudo-Go):

```go
func hasForbiddenNodes(untrustedXML somepkg.XML) bool {
	return hasComments(untrustedXML) || hasDirectives(untrustedXML) || hasProcessingInstructions(untrustedXML)
}

func verifyDSIG(untrusted string, publicKey interface{}) (Assertion, error) {
	untrustedXMLDOM, err := somepkg.ParseXML(untrusted)
	if nil != err {
		return nil, err
	}
	if reserializeXML(untrustedXMLDOM) != untrusted {
		return nil, errors.New("Bug in somepkg: XML didn’t round-trip successfully")
	}
	if hasForbiddenNodes(untrustedXMLDOM) {
		return nil, errors.New("Forbidden node in XML")
	}
	untrustedSignature, err := extractSignature(untrustedXMLDOM)
	if nil != err {
		return nil, err
	}
	if nil == untrustedSignature {
		panic("nil return from extractSignature with no error")
	}
	if untrustedSignature.Parent != untrustedXMLDOM.Root {
		return nil, errors.New("Parent of the signature isn’t the toplevel node")
	}
	untrustedAssertion, err := extractAssertion(untrustedXMLDOM)
	if nil != err {
		return nil, errors.New("Could not extract assertion")
	}
	var untrustedAssertionBytes []byte = untrustedAssertion.Serialize()
	if err = verifySignature(untrustedAssertionBytes, untrustedSignature, publicKey); nil != err {
		return err
	}

	// At last, something signed!
	signedAssertion := untrustedAssertionBytes
	reparsed, err := somepkg.Parse(signedAssertion)
	if nil != err {
		return nil, err
	}

	// Optional, though I recommend it
	if !deepEqual(reparsed, untrustedAssertion) {
		return nil, errors.New("somepkg bug: assertion didn’t round trip!")
	}

	return reparsed, nil
}
```

One also needs to handle all of the parts of SAML that restrict when an assertion is valid.

**Comment by krestivo-altoslabs** [comment]
7 years ago, someone asked for SAML. And now, yet another will ask for it.

**Comment by aphorise** [comment]
Very valid points @DemiMarie.

There are solutions via WS-Federation and other providers (ADFS / AAD) in this space that facilitate for IdP transitions of these older protocols in environments or scenarios that they were original envisaged in (cloud or mix of on-prem intranet with internet). Leveraging these to get obtain OIDC / JWT would seem for me to the the easiest and most possible approach rather than hoping for a support for SAML.

The Oath2 part of this original request may still be relevant today over and above that of SAML IMO.

**Comment by jd-sandk** [comment]
+1 for SAML support.

**Comment by gautaz** [comment]
[SATOSA](https://github.com/IdentityPython/SATOSA) can act as a [proxy from OIDC to SAMLv2](https://github.com/IdentityPython/SATOSA/blob/master/doc/oidc-to-saml2.md).

Has anyone tried this with Vault ?

**Comment by JubbaSmail** [comment]
@armon @jefferai @radeksimko any plan to support SAML2.0 in 2023?

**Comment by jd-sandk** [comment]
If it helps anyone get past the pain point we were having . . .

We started using a GitHub Auth mount in the Vault server via Terraform (`vault_github_auth_backend`), coupled with `vault_github_team` for each team and very well defined Policies: `vault_policy` for each of the teams.

(Now we can sign-in to Vault with our GitHub Personal Access Tokens and everything works as intended and is locked down as needed) [the ability to use Vault is based on our GitHub Teams's membership, so everything is basically automated now in-regards to the vault on-boarding/off-boarding process, as long as the teams do that correctly correctly on their end; I would still recommend Auditing the GitHub Teams every so often tho . . . incase you find stragglers from incorrectly deactivated employees]

**Comment by maxb** [comment]
> * The first one, as you mentioned, is off-boarding the employee the moment the HR team disable the account.

Are you potentially referring to some optional extension of SAML here? AFAIK regardless of whether the user logs to a system such as Vault using OIDC or SAML, there is no mechanism to then terminate their Vault session if their account is disabled in the identity provider.

----

> * Centralized login to help in account recovery, control over the minimum security measure (ex 2nd factor, device management, region, VPN, etc..) and centralized audit logs.
> * Vault Role and Permissions based on SAML attributes (Organization Unit, Group, etc)
> * Listing Vault Apps in "App Launcher" based on access, So employees will see the list of Vaults instances that can be accessed
> 
> OIDC Auth also only covers some attributes. For example, it's only possible to get Google Groups and not Organization Unit.

I understand these are genuine issues for you, but all of them can be supported via OIDC (speaking conceptually, in terms of the protocol). It sounds like you have an Identity Provider which has better support for SAML than OIDC?

**Comment by JubbaSmail** [comment]
@maxb, So with SAML + SCIM the account will be disabled in Vault once it's disabled in the Identity Provider. 
Plus the Identity Provider audit logs don't display OCID activity,
Also, OCID does not support MDM Policy nor Region or IP policy,
And OCID can't be listed in the SAML Apps list that will display the apps based on access,
Finally, OCID can't read Organization Unit or other User attributes from the Identity Provider.

In conclusion, OCID is limited in comparison to SAML.

**Comment by maxb** [comment]
> So with SAML + SCIM the account will be disabled in Vault once it's disabled in the Identity Provider.

Yes, but SCIM is an additional technology separate to SAML, which could theoretically be combined with OIDC just as well as with SAML. I do appreciate this does not help you much right now, but it is good to ensure the facts are clearly stated so everyone is on the same page.



> @maxb, So with SAML + SCIM the account will be disabled in Vault once it's disabled in the Identity Provider. Plus the Identity Provider audit logs don't display OCID activity, Also, OCID does not support MDM Policy nor Region or IP policy, And OCID can't be listed in the SAML Apps list that will display the apps based on access, Finally, OCID can't read Organization Unit or other User attributes from the Identity Provider.
> 
> In conclusion, OCID is limited in comparison to SAML.

No... in conclusion, your identity provider's implementation of OIDC is limited in comparison to your identity provider's implementation of SAML.

I realise this is not the **most** helpful thing to be saying, but it does highlight that there is another potential way forward - approaching your identity provider vendor about them rectifying these feature gaps.

**Comment by JubbaSmail** [comment]
@maxb  It's not the Identity Provider limitation; it's the protocol limitation (OIDC). In my case, we are using Google Workspace SAML 2.0.

In OIDC implementation, the login process can only be initiated by the Service Provider (SP), never by the Identity Provider (IdP). The IdP can't control who can access this SP, so from the IdP side, everyone in the domain can log in, and it's the SP's job to filter the access.

In SAML 2.0, IdP can block users from accessing SP and present only accessible Apps to the users. See the example below:
![image](https://user-images.githubusercontent.com/2150143/219956519-6c40d5ab-7497-4dad-81c6-2646cb139f27.png)

So for my organization's SAML 2.0 Apps, I can define extra security policies to access the App (Vault), for Example: enforcing MDM Policy (Device Strong Password, Disk Encryption, Minimum OS version, Minimum Browser Version, Company-Owned Device) or Network Policy (VPN, Region).
More details about Context-Aware Access in Google Workspace: https://support.google.com/a/answer/12645308?hl=en

So in my organization, there are over 60 SAML Apps, all defined and organized in one centralized place, with a very easy way to onboard/offboard employees or request/approve access to certain apps based on Organization Unit/Group. Vault is the only App in my organization that is not supporting SAML 2.0, missing our all advanced security setup in the organisation.

**Comment by remilapeyre** [comment]
Hi @JubbaSmail, I am currently working on a SAML authentication method that is working. I will share it in a couple of days.

**Comment by bds-congnguyen** [comment]
looking for SAML support :D

**Comment by DemiMarie** [comment]
> Hi @JubbaSmail, I am currently working on a SAML authentication method that is working. I will share it in a couple of days.

Do you use the approach I mentioned in https://github.com/hashicorp/vault/issues/370#issuecomment-974911541?

**Comment by aphorise** [comment]
Seems like SAML authentication is in the next enterprise release of Vault (1.15.0) as per the recent [CHANGELOG.md](https://github.com/hashicorp/vault/blob/main/CHANGELOG.md) **1.15-rc1** entries:

>1.15.0-rc1 (September 13, 2023)
> FEATURES:
> SAML Auth Method (enterprise): Enable users to authenticate with Vault using their identity in a SAML Identity Provider.

To be honest - I dont believe this would invalidate the requests made herein and it may be best if any community / OSS offered plugin that may be developed is called `saml-oss-auth` to avoid any confusion or clash. What's more it would be interesting if the community edition out performs or offers more than what will eventually be offered on enterprise versions only 😄

**Comment by solutiongeek** [comment]
The BSL license change is limited to our core products. It does not affect any of our plugins, which remain licensed under MPL, and has no impact on providers created by third parties.

**Comment by DemiMarie** [comment]
@solutiongeek Nit: Business Source License is BUSL, BSL is Boost Software License which is very different.

**Comment by austingebauer** [comment]
Closing this issue as completed. A [SAML auth method](https://developer.hashicorp.com/vault/docs/auth/saml) has been added to Vault Enterprise in 1.15.
